Fix Part of #13162: Add argument schema for ProfilePictureHandler and ProfilePictureHandlerByUsernameHandler classes

mahendra1290 opened the pull request:
Add argument schema for ```ProfilePictureHandler``` and ```ProfilePictureHandlerByUsernameHandler```.

## Overview
<!--
READ ME FIRST:
Please answer *both* questions below and check off every point from the Essential Checklist!
If there is no corresponding issue number, fill in N/A where it says [fill_in_number_here] below in 1.
-->

1. This PR fixes or fixes part of  #13162.
2. This PR does the following: This PR add argument schema for ```ProfilePictureHandler``` and ```ProfilePictureHandlerByUsernameHandler```.

## Essential Checklist

- [x] The PR title starts with "Fix #bugnum: ", followed by a short, clear summary of the changes. (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [x] The linter/Karma pre-submit checks have passed locally on your machine.
- [x] "Allow edits from maintainers" is checked. (See [here](https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/allowing-changes-to-a-pull-request-branch-created-from-a-fork) for instructions on how to enable it.)
  - This lets reviewers restart your CircleCI tests for you.
- [x] The PR is made from a branch that's **not** called "develop".

## Proof that changes are correct
![Screenshot from 2021-08-30 14-45-58](https://user-images.githubusercontent.com/42772952/131326554-1895bfeb-5f46-46da-a3d3-bce9029e0544.png)

![Screenshot from 2021-08-30 14-47-24](https://user-images.githubusercontent.com/42772952/131326594-e96c8afa-cf15-45bf-9140-4b32aa2eed51.png)


![Screenshot from 2021-08-30 14-16-22](https://user-images.githubusercontent.com/42772952/131313695-9bbf574f-70df-4006-a168-c812c719f1ea.png)

<!--
Add videos/screenshots of the user-facing interface in various display sizes (mainly phone, tablet, and desktop display size) to demonstrate that the changes made in this PR work correctly.
If the changes in your PRs are autogenerated via a script and you cannot provide proof for the changes then please leave a comment "No proof of changes needed because {{Reason}}".
-->

## PR Pointers

- Make sure to follow the [instructions for making a code change](https://github.com/oppia/oppia/wiki/Contributing-code-to-Oppia#instructions-for-making-a-code-change).
- Oppiabot will notify you when you don't add a PR_CHANGELOG label. If you are unable to do so, please @-mention a code owner (who will be in the Reviewers list), or ask on [Gitter](https://gitter.im/oppia/oppia-chat).
- For what code owners will expect, see the [Code Owner's wiki page](https://github.com/oppia/oppia/wiki/Oppia%27s-code-owners-and-checks-to-be-carried-out-by-developers).
- Make sure your PR follows conventions in the [style guide](https://github.com/oppia/oppia/wiki/Coding-style-guide), otherwise this will lead to review delays.
- Never force push. If you do, your PR will be closed.
- Oppiabot can assign anyone for review/help if you leave a comment like the following: "{{Question/comment}} @{{reviewer_username}} PTAL"
- Some of the e2e tests are flaky and can fail for reasons unrelated to your PR. We are working on fixing this, but in the meantime, if you need to restart the tests, please check the ["If your build fails" wiki page](https://github.com/oppia/oppia/wiki/If-your-build-fails).

mahendra1290 committed:
Add schema for handler class arguments

added argument schema for ProfilePictureHandler and ProfilePictureHandlerByUsernameHandler.

Nik-09 commented on core/controllers/profile.py:
```diff
@@ -237,6 +237,12 @@ class ProfilePictureHandler(base.BaseHandler):
     """
 
+    URL_PATH_ARGS_SCHEMAS = {}
+
```
Remove this extra new line.

mahendra1290 replied:
@Nik-09 okay removing it. Should I add screenshots in PR or here in the comments?

mahendra1290 replied:
@Nik-09 should I  remove line 240 or just an extra new line below it?

Nik-09 replied:
In the PR description -- as a proof that changes are correct.

Nik-09 replied:
> should I remove line 240 or just an extra new line below it?

Only remove the new line.

Nik-09 reviewed (commented):
Added one comment.
Also as a fun part -- have you tested this handler from frontend, if so please provide some screenshots.
Thanks!

Nik-09 commented:
@mahendra1290 The backend tests are failing -- please read the instruction in the provided in the issue #13162. You have to remove the handler name from the list in payload_validator.py file.

mahendra1290 committed:
remove extra new line

mahendra1290 commented:
> @mahendra1290 The backend tests are failing -- please read the instruction in the provided in the issue #13162. You have to remove the handler name from the list in payload_validator.py file.

The link to the wiki is not redirecting to the correct page it's redirecting to the home page of the wiki.

mahendra1290 committed:
remove handler class name

Nik-09 commented:
> The link to the wiki is not redirecting to the correct page it's redirecting to the home page of the wiki.

Thank you for pointing this, updated!

mahendra1290 commented:
> @mahendra1290 The backend tests are failing -- please read the instruction in the provided in the issue #13162. You have to remove the handler name from the list in payload_validator.py file.

I have removed the handler name from payload_validators.py file

Nik-09 commented on core/controllers/profile.py:
```diff
@@ -256,6 +261,18 @@ class ProfilePictureHandlerByUsernameHandler(base.BaseHandler):
+            }
+        }
+    }
+
```
Remove this line also.

mahendra1290 replied:
@Nik-09 done.

Nik-09 reviewed (approved):
One last nit -- Just remove one line.

mahendra1290 committed:
remove empty line

rohitkatlaa reviewed (approved):
LGTM. Thanks!

vojtechjelinek commented on core/controllers/profile.py:
```diff
@@ -256,6 +261,17 @@ class ProfilePictureHandlerByUsernameHandler(base.BaseHandler):
 
+    URL_PATH_ARGS_SCHEMAS = {
+        'username': {
+            'schema': {
+                'type': 'basestring'
+            }
```
Can we introduce more rules for username? I think there exist a limit for minimal/maximal length of the username and also the allowed characters.

mahendra1290 replied:
Hi, @vojtechjelinek after searching a lot into the project I got these validations.
1. Max length validation of 30 char
2. Only alphanumeric ['^[A-Za-z0-9]+$'] characters are allowed.
3. Should not be system reserved user name like oppia, admin.
![Screenshot from 2021-08-31 01-03-35](https://user-images.githubusercontent.com/42772952/131394767-2ead83b4-4038-42c2-92dd-9cadc0cf9f1d.png)

mahendra1290 replied:
@vojtechjelinek I think we need to add validators right ?

mahendra1290 replied:
@vojtechjelinek In other places we're using same username schema we need to add rules everywhere if we'll be updating it here.

vojtechjelinek replied:
Yep

mahendra1290 replied:
@vojtechjelinek username validation is there when signing up and if in URL invalid username is passed no data will be returned.

vojtechjelinek replied:
True, but I think at least the basic validation of length and allowed chars (regex) should be done here too.

mahendra1290 replied:
Hi, @vojtechjelinek I have added this method locally in `schema_utils.py`.

![Screenshot from 2021-08-31 16-44-36](https://user-images.githubusercontent.com/42772952/131493125-1d3b695a-99d6-47bd-b857-e5d485be5e96.png)


and in `profile.py`.

![Screenshot from 2021-08-31 16-43-58](https://user-images.githubusercontent.com/42772952/131493063-6f0a03ae-f205-4dd1-a80b-f6150f5c1efe.png)

Let me know if it is correct or not and what should I do next?

mahendra1290 replied:
@vojtechjelinek There is also repetition of code more importantly the **last part of checking system reserved name** can we add this validation somewhere else so that we can resuse it.

vojtechjelinek replied:
Hmm, I might have some ideas, feel free to push the code so that I can give you comments into the code itself.

mahendra1290 replied:
@vojtechjelinek okay

mahendra1290 replied:
just pushed changes, have a look and let me know what I can do.

vojtechjelinek replied:
Hey, thinking about it I think we can import user_domain.py inside the schema_utils.py and then use the `user_domain.UserSettings.require_valid_username` in a try-except block, when the error is catched return `False` otherwise `True`. That way we can get rid of the duplication.

mahendra1290 replied:
Ya we can do that, I'll try and push it

mahendra1290 replied:
I am getting this error during git push

```
Could not automatically determine credentials. Please set GOOGLE_APPLICATION_CREDENTIALS or explicitly create credentials and re-run the application. For more information, please see https://cloud.google.com/docs/authentication/getting-started
09:51:43 ERROR Third Party Python Lint Checks Manager: 6.8 secs

Unable to run the complete lint test, please check the following stack trace and fix the errors:
+--------------------------+
Traceback (most recent call last):
  File "/home/maahi/open-source/oppia/scripts/concurrent_task_utils.py", line 101, in run
    self.task_results = self.func()
  File "/home/maahi/open-source/oppia/scripts/linters/python_linter.py", line 255, in perform_all_lint_checks
    linter_stdout.append(self.lint_py_files())
  File "/home/maahi/open-source/oppia/scripts/linters/python_linter.py", line 123, in lint_py_files
    exit=False
  File "/home/maahi/open-source/oppia/../oppia_tools/pylint-quotes-0.1.9/pylint/lint/run.py", line 330, in __init__
    linter.load_plugin_modules(plugins)
  File "/home/maahi/open-source/oppia/../oppia_tools/pylint-quotes-0.1.9/pylint/lint/pylinter.py", line 540, in load_plugin_modules
    module = astroid.modutils.load_module_from_name(modname)
  File "/home/maahi/open-source/oppia/../oppia_tools/pylint-quotes-0.1.9/astroid/modutils.py", line 216, in load_module_from_name
    return importlib.import_module(dotted_name)
  File "/home/maahi/.pyenv/versions/3.7.10/lib/python3.7/importlib/__init__.py", line 127, in import_module
    return _bootstrap._gcd_import(name[level:], package, level)
  File "<frozen importlib._bootstrap>", line 1006, in _gcd_import
  File "<frozen importlib._bootstrap>", line 983, in _find_and_load
  File "<frozen importlib._bootstrap>", line 967, in _find_and_load_unlocked
  File "<frozen importlib._bootstrap>", line 677, in _load_unlocked
  File "<frozen importlib._bootstrap_external>", line 728, in exec_module
  File "<frozen importlib._bootstrap>", line 219, in _call_with_frames_removed
  File "/home/maahi/open-source/oppia/scripts/linters/pylint_extensions.py", line 30, in <module>
    from core.controllers import payload_validator
  File "/home/maahi/open-source/oppia/core/controllers/payload_validator.py", line 25, in <module>
    import schema_utils
  File "/home/maahi/open-source/oppia/schema_utils.py", line 36, in <module>
    from core.domain.user_domain import UserSettings
  File "/home/maahi/open-source/oppia/core/domain/user_domain.py", line 31, in <module>
    (user_models,) = models.Registry.import_models([models.NAMES.user])
  File "/home/maahi/open-source/oppia/core/platform/models.py", line 423, in import_models
    return cls._get().import_models(model_names)
  File "/home/maahi/open-source/oppia/core/platform/models.py", line 187, in import_models
    from core.storage.user import gae_models as user_models
  File "/home/maahi/open-source/oppia/core/storage/user/gae_models.py", line 38, in <module>
    (base_models,) = models.Registry.import_models([models.NAMES.base_model])
  File "/home/maahi/open-source/oppia/core/platform/models.py", line 423, in import_models
    return cls._get().import_models(model_names)
  File "/home/maahi/open-source/oppia/core/platform/models.py", line 116, in import_models
    from core.storage.base_model import gae_models as base_model
  File "/home/maahi/open-source/oppia/core/storage/base_model/gae_models.py", line 52, in <module>
    transaction_services = models.Registry.import_transaction_services()
  File "/home/maahi/open-source/oppia/core/platform/models.py", line 476, in import_transaction_services
    return cls._get().import_transaction_services()
  File "/home/maahi/open-source/oppia/core/platform/models.py", line 249, in import_transaction_services
    from core.platform.transactions import cloud_transaction_services
  File "/home/maahi/open-source/oppia/core/platform/transactions/cloud_transaction_services.py", line 28, in <module>
    CLIENT = datastore.Client()
  File "/home/maahi/open-source/oppia/third_party/python_libs/google/cloud/datastore/client.py", line 288, in __init__
    _http=_http,
  File "/home/maahi/open-source/oppia/third_party/python_libs/google/cloud/client.py", line 277, in __init__
    _ClientProjectMixin.__init__(self, project=project, credentials=credentials)
  File "/home/maahi/open-source/oppia/third_party/python_libs/google/cloud/client.py", line 225, in __init__
    project = self._determine_default(project)
  File "/home/maahi/open-source/oppia/third_party/python_libs/google/cloud/datastore/client.py", line 317, in _determine_default
    return _determine_default_project(project)
  File "/home/maahi/open-source/oppia/third_party/python_libs/google/cloud/datastore/client.py", line 93, in _determine_default_project
    project = _base_default_project(project=project)
  File "/home/maahi/open-source/oppia/third_party/python_libs/google/cloud/_helpers.py", line 186, in _determine_default_project
    _, project = google.auth.default()
  File "/home/maahi/open-source/oppia/third_party/python_libs/google/auth/_default.py", line 488, in default
    raise exceptions.DefaultCredentialsError(_HELP_MESSAGE)
google.auth.exceptions.DefaultCredentialsError: Could not automatically determine credentials. Please set GOOGLE_APPLICATION_CREDENTIALS or explicitly create credentials and re-run the application. For more information, please see https://cloud.google.com/docs/authentication/getting-started

--------------------------------------------------

--------------------------------------------------
Some of the linting functions may not run until the above errors gets fixed
----------------------------------------
Please fix the errors below:
----------------------------------------
---------------------------
Checks Not Passed.
---------------------------
Push failed, please correct the linting issues above.
Already on 'handler-class-argument-schema'```

vojtechjelinek replied:
Just remove the `(user_models,) = models.Registry.import_models([models.NAMES.user])` from _user_domain.py_ as it is not needed there and it should fix this error.

mahendra1290 replied:
It worked Thanks.

mahendra1290 replied:
I have pushed changes please let me know what I should do next.

vojtechjelinek reviewed (changes requested):
Thanks! Left one comment.

mahendra1290 committed:
Add username validation

add validation function for username in schema_utils.py and using the
same for extra validation in handler schema arguments.

mahendra1290 committed:
use user_domain.UserSettings.require_valid_username for username validation

mahendra1290 committed:
remove unused line

mahendra1290 committed:
fix lint errors

vojtechjelinek reviewed (approved):
LGTM!

mahendra1290 committed:
add unit test for is_valid_username_string method

mahendra1290 commented:
@vojtechjelinek I have also added tests, please have a look, when this PR will be merged?

vojtechjelinek commented:
@mahendra1290 You still need a review from @seanlip and @DubeySandeep, then after all tests pass it will be merged.

seanlip reviewed (approved):
Sorry for the delay -- LGTM for codeowner file!

mahendra1290 commented:
@DubeySandeep please review this PR and let me know about any further changes I need to do.

mahendra1290 commented:
@DubeySandeep PTAL!

mahendra1290 commented:
@vojtechjelinek It's showing a stale build should I merge develop into this branch and then commit?

mahendra1290 committed:
Merge branch 'oppia:develop' into handler-class-argument-schema

mahendra1290 commented:
Hi @kevintab95 Please review this PR and let me know if any changes are needed.

mahendra1290 committed:
Merge remote-tracking branch 'upstream/develop' into handler-class-argument-schema

kevintab95 commented on core/controllers/payload_validator.py:
```diff
@@ -36,7 +36,6 @@ def validate(
         allowed_extra_args: bool,
         allow_string_to_bool_conversion: bool = False
 ) -> Tuple[Dict[str, Any], List[str]]:
-
```
Is this an expected change in this PR i.e. why remove this line? Ditto below.

mahendra1290 replied:
I think it was removed during auto-formatting there is no space between the function name and its doc string at other places. I can add it back if needed

kevintab95 reviewed (approved):
@mahendra1290 just one nit. Please address it before merging. I am approving to unblock my review hold. Thanks!

mahendra1290 committed:
Merge remote-tracking branch 'upstream/develop' into handler-class-argument-schema

Update with latest develop branch

mahendra1290 commented:
hi @vojtechjelinek Please let me know if any further changes are needed

Nik-09 commented:
@mahendra1290 Please update the title of the pr as **Fix Part of #<issue_number>:** 
Thanks!
